wallet/wallet.go

```diff
@@ -4709,6 +4691,108 @@ func (s sigDataSource) GetKey(a stdaddr.Address) ([]byte, dcrec.SignatureType, b
 }
 func (s sigDataSource) GetScript(a stdaddr.Address) ([]byte, error) { return s.script(a) }
 
+// CreateVspPayment receives a tx and ensures that it pays the correct fee
+// amount to the correct address. It then signs that tx and adds it to the
+// wallet without broadcasting it to the network.
+func (w *Wallet) CreateVspPayment(ctx context.Context, tx *wire.MsgTx, fee dcrutil.Amount,
+	feeAddr stdaddr.Address, feeAcct uint32, changeAcct uint32) error {
+
+	// Reserve new outputs to pay the fee if outputs have not already been
+	// reserved.  This will be the case for fee payments that were begun on
+	// already purchased tickets, where the caller did not ensure that fee
+	// outputs would already be reserved.
+	if len(tx.TxIn) == 0 {
+		const minconf = 1
+		inputs, err := w.ReserveOutputsForAmount(ctx, feeAcct, fee, minconf)
+		if err != nil {
+			return fmt.Errorf("unable to reserve outputs: %w", err)
+		}
+		for _, in := range inputs {
+			tx.AddTxIn(wire.NewTxIn(&in.OutPoint, in.PrevOut.Value, nil))
+		}
+		// The transaction will be added to the wallet in an unpublished
+		// state, so there is no need to leave the outputs locked.
+		defer func() {
+			for _, in := range inputs {
+				w.UnlockOutpoint(&in.OutPoint.Hash, in.OutPoint.Index)
+			}
+		}()
+	}
+
+	var input int64
+	for _, in := range tx.TxIn {
+		input += in.ValueIn
+	}
+	if input < int64(fee) {
+		err := fmt.Errorf("not enough input value to pay fee: %v < %v",
+			dcrutil.Amount(input), fee)
+		return err
+	}
+
+	vers, feeScript := feeAddr.PaymentScript()
+
+	addr, err := w.NewChangeAddress(ctx, changeAcct)
+	if err != nil {
+		log.Warnf("failed to get new change address: %v", err)
+		return err
+	}
+	var changeOut *wire.TxOut
+	switch addr := addr.(type) {
+	case Address:
+		vers, script := addr.PaymentScript()
+		changeOut = &wire.TxOut{PkScript: script, Version: vers}
+	default:
+		return fmt.Errorf("failed to convert '%T' to wallet.Address", addr)
+	}
+
+	tx.TxOut = append(tx.TxOut[:0], &wire.TxOut{
+		Value:    int64(fee),
+		Version:  vers,
+		PkScript: feeScript,
+	})
+	feeRate := w.RelayFee()
+	scriptSizes := make([]int, len(tx.TxIn))
+	for i := range scriptSizes {
+		scriptSizes[i] = txsizes.RedeemP2PKHSigScriptSize
+	}
+	est := txsizes.EstimateSerializeSize(scriptSizes, tx.TxOut, txsizes.P2PKHPkScriptSize)
+	change := input
+	change -= tx.TxOut[0].Value
+	change -= int64(txrules.FeeForSerializeSize(feeRate, est))
+	if !txrules.IsDustAmount(dcrutil.Amount(change), txsizes.P2PKHPkScriptSize, feeRate) {
+		changeOut.Value = change
+		tx.TxOut = append(tx.TxOut, changeOut)
+		txauthor.RandomizeOutputPosition(tx.TxOut, 0)
```

Comment:
Please review this usage of `txauthor.RandomizeOutputPosition`, unsure if I've done this properly

Reply:
Done